operator/src/lib.rs

```diff
@@ -207,23 +220,23 @@ impl SparkState {
 
     /// Reconcile the cluster according to provided spec. Start with master nodes and continue to worker and history-server nodes.
     /// Create missing pods or delete excess pods to match the spec.
-    pub async fn reconcile_cluster(&self) -> SparkReconcileResult {
-        trace!("Starting {} reconciliation", SparkNodeType::Master.as_str());
-
-        if let Some(node_info) = &self.node_information {
-            self.reconcile_node(&SparkNodeType::Master, &node_info.master)
-                .await?;
-            self.reconcile_node(&SparkNodeType::Worker, &node_info.worker)
-                .await?;
-
-            if let Some(history_server) = &node_info.history_server {
-                self.reconcile_node(&SparkNodeType::HistoryServer, history_server)
-                    .await?;
-            } else {
-                return Ok(ReconcileFunctionAction::Continue);
+    pub async fn reconcile_cluster(&self, node_type: &SparkNodeType) -> SparkReconcileResult {
+        trace!("Starting {} reconciliation", node_type.as_str());
+
+        if let Some(pod_info) = &self.pod_information {
+            match node_type {
+                SparkNodeType::Master => {
+                    return self.reconcile_node(node_type, &pod_info.master).await
+                }
+                SparkNodeType::Worker => {
+                    return self.reconcile_node(node_type, &pod_info.worker).await
+                }
+                SparkNodeType::HistoryServer => {
+                    if let Some(history_server) = &pod_info.history_server {
+                        return self.reconcile_node(node_type, history_server).await;
+                    }
+                }
             }
-        } else {
-            return Ok(ReconcileFunctionAction::Done);
         }
 
         Ok(ReconcileFunctionAction::Continue)
```

Comment:
If you like you can change the whole method to

```rust
    pub async fn reconcile_cluster(&self, node_type: &SparkNodeType) -> SparkReconcileResult {
        trace!("Starting {} reconciliation", node_type.as_str());

        match (node_type, &self.pod_information) {
            (SparkNodeType::Master, Some(pod_info)) => {
                self.reconcile_node(node_type, &pod_info.master).await
            }
            (SparkNodeType::Worker, Some(pod_info)) => {
                self.reconcile_node(node_type, &pod_info.worker).await
            }
            (
                SparkNodeType::HistoryServer,
                Some(PodInformation {
                    history_server: Some(history_server),
                    ..
                }),
            ) => self.reconcile_node(node_type, history_server).await,
            _ => Ok(ReconcileFunctionAction::Continue),
        }
    }
```

Not necessarilty easier to read. I sometimes try to rewrite things just to understand Rust better. I'll leave it upt o you what you want to do.